### DISCUSSION TITLE
Translation

### DISCUSSION THREAD

**Discussion by ichnilatis-gr** [koreader/koreader#10774] [Q&A]
Could anybody please explain to me what "mu" stands for [here](https://hosted.weblate.org/translate/koreader/koreader/en/?checksum=df3db57dfc73cdf7)? 

Thanks!

**Comment by mergen3107** [comment]
This supposed to be “ x m y s”, where m is “minutes” and s is “seconds”, x and y their values. \u{2009} is supposed to be a Univode character that represents thin space: https://www.compart.com/en/unicode/U+2009

This is all I got from the code line: https://github.com/koreader/koreader/blob/master/frontend/datetime.lua#L153

**Comment by mergen3107** [comment]
I left a comment on Weblate, I hope it’s correct :D

**Reply by poire-z** [reply]
I think it's an issue if the `\` is lost in the translation (only in the display of the original string ? but so also in the translated string as user won't add the \ back, or may lose it if they edit the whole thing). @Frenzie ?

**Reply by Frenzie** [reply]
https://github.com/koreader/koreader-translations/commit/34fbe88f267a244ce1289692ee8b60dda777d460

It looks like the `\` wasn't picked up by xgettext, which smells like a bug. I'd expect either simply `\u` or otherwise `\\u`. The first course of action should be to check it against xgettext 0.22 but I'll have to compile it from scratch first. The [release notes](https://lists.gnu.org/archive/html/info-gnu/2023-06/msg00003.html) do happen to mention something about `\u`  in relation to Tcl.

**Reply by Frenzie** [reply]
Alas, no difference.
```
$ /home/frans/src/ko/gettext-0.22/gettext-tools/src/xgettext --version
xgettext (GNU gettext-tools) 0.22
Copyright (C) 1995-2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Written by Ulrich Drepper.
$ /home/frans/src/ko/gettext-0.22/gettext-tools/src/xgettext --from-code=utf-8 \
        --keyword=C_:1c,2 --keyword=N_:1,2 --keyword=NC_:1c,2,3 \
        --add-comments=@translators \
        reader.lua `find frontend -iname "*.lua" | sort` \
        `find plugins -iname "*.lua" | sort` \
        `find tools -iname "*.lua" | sort` \
        -o l10n/templates/koreader.pot
```

**Reply by Frenzie** [reply]
@poire-z The utf-8 codepoint may be the better approach regardless though? Iirc that's much friendlier than whatever `\u` etc. might do.

**Reply by NiLuJe** [reply]
The naked codepoint makes sense when we can actually *see* it and it's somewhat standard enough that we *can* actually see it ;p; which is rarely the case for what we do, so I tend to prefer explicit unicode escapes like what @poire-z just did.

On the other hand, that is an unfortunate side-effect ;o).

FWIW, escaping appears to behave in C, because I'm doing that in [KoboUSBMS](https://github.com/koreader/KoboUSBMS/blob/3daab316d3aff2b43ced9c0b18e6ecdeec953e4a/usbms.c#L1185-L1186), and they've been (mostly ;p) left alone...

**Reply by Frenzie** [reply]
I posted a diff earlier (and deleted since it was an unfinished though) that I thought wouldn't make sense to upstream, but actually it would. So the solution's probably fairly simple.
<https://www.lua.org/manual/5.3/manual.html#3.1>
> The UTF-8 encoding of a Unicode character can be inserted in a literal string with the escape sequence \u{XXX} (note the mandatory enclosing brackets), where XXX is a sequence of one or more hexadecimal digits representing the character code point. 

--

> The naked codepoint makes sense when we can actually see it and it's somewhat standard enough that we can actually see it ;p; which is rarely the case for what we do, so I tend to prefer explicit unicode escapes like what @poire-z just did.

That's precisely the reason I said it's probably better to use the codepoint. The translator simply *sees* the character rather than some programmer mumbo jumbo.

Edit: but maybe the same will happen with `\u`. Needs testing.

**Reply by poire-z** [reply]
`\u{2345}` is much friendlier when looking at/writing the code.
But if there's an issue with translation, you could revert only those part of a translatable string to use \x bytes.
https://github.com/koreader/koreader/commit/626864f8561db8c8ec18bebd71a0f7c6da8fc68e
It looks like there's only this one mentionned above:
```diff
-                    return T(C_("Time", "%1m\xE2\x80\x89%2s"), "0", string.format("%d", seconds))
+                    return T(C_("Time", "%1m\u{2009}%2s"), "0", string.format("%d", seconds)) -- use a thin space
```

**Reply by Frenzie** [reply]
@poire-z Looking at the xgettext code in closer detail it has special handling for `\x`, so the same special handling should be added for `\u`, which will then have an identical result. But that particular string should preferably be reverted to `\x` for the moment to avoid generating incorrect translations and wasting people's time.

```c
                    case 'x':
                      {
                        int num = 0;
                        int i = 0;

                        for (i = 0; i < 2; i++)
                          {
                            c = phase1_getc ();
                            if (c >= '0' && c <= '9')
                              num += c - '0';
                            else if (c >= 'a' && c <= 'f')
                              num += c - 'a' + 10;
                            else if (c >= 'A' && c <= 'F')
                              num += c - 'A' + 10;
                            else
                              {
                                phase1_ungetc (c);
                                break;
                              }

                            if (i == 0)
                              num *= 16;
                          }

                        if (i == 2)
                          string_add (num);
                      }

                      break;
```

**Reply by Frenzie** [reply]
@poire-z But wait, is the escape sequence you used different from the one in Lua 5.3? For reference, here's a quick POC that seems to be working (give or take potential edge cases)… well, based on that phrase `the escape sequence \u{XXX} (note the mandatory enclosing brackets), where XXX is a sequence of one or more hexadecimal digits representing the character code point`.

```diff
diff --git a/gettext-tools/src/x-lua.c b/gettext-tools/src/x-lua.c
index 78f0d179b..80ef8c619 100644
--- a/gettext-tools/src/x-lua.c
+++ b/gettext-tools/src/x-lua.c
@@ -634,6 +634,51 @@ phase3_get (token_ty *tp)
                       break;
                     case 't':
                       string_add ('\t');
+                      break;
+                    case 'u':
+                      {
+                        int num = 0;
+                        int i = 0;
+
+                        c = phase1_getc ();
+
+                        if (c == '{')
+                          {
+                            for (i = 0; i < 4; i++)
+                              {
+                                c = phase1_getc ();
+                                if (c >= '0' && c <= '9')
+                                  num += c - '0';
+                                else if (c >= 'a' && c <= 'f')
+                                  num += c - 'a' + 10;
+                                else if (c >= 'A' && c <= 'F')
+                                  num += c - 'A' + 10;
+                                else
+                                  {
+                                    phase1_ungetc (c);
+                                    break;
+                                  }
+
+                                if (i == 0)
+                                  num *= 16;
+                              }
+
+                            if (i == 4)
+                              {
+                                c = phase1_getc ();
+                                if (c == '}')
+                                  string_add (num);
+                                else
+                                  {
+                                    phase1_ungetc (c);
+                                    break;
+                                  }
+                              }
+                          }
+                        else
+                          phase1_ungetc (c);
+                      }
+
                       break;
                     case 'v':
                       string_add ('\v');
```

Edit: though I realize now I somewhat misunderstood what that line in the reference was saying, so this doesn't make much sense really. xD

**Reply by Frenzie** [reply]
So for us, this would be what we want.

```diff
diff --git a/gettext-tools/src/x-lua.c b/gettext-tools/src/x-lua.c
index 78f0d179b..5d908444f 100644
--- a/gettext-tools/src/x-lua.c
+++ b/gettext-tools/src/x-lua.c
@@ -635,6 +635,10 @@ phase3_get (token_ty *tp)
                     case 't':
                       string_add ('\t');
                       break;
+                    case 'u':
+                      string_add ('\\');
+                      string_add (c);
+                      break;
                     case 'v':
                       string_add ('\v');
                       break;
```

I suspect this means the following current strings are messed up?

```diff
@@ -2614,11 +2614,11 @@ msgid "Fantasy"
 msgstr ""
 
 #: frontend/apps/reader/modules/readerfont.lua:490
-msgid "Emoji u{1F60A}"
+msgid "Emoji \\u{1F60A}"
 msgstr ""
 
 #: frontend/apps/reader/modules/readerfont.lua:491
-msgid "Fang Song u{4EFF}u{5B8B}"
+msgid "Fang Song \\u{4EFF}\\u{5B8B}"
 msgstr ""
```

**Reply by poire-z** [reply]
Yes, I think I've read recently that `\u{...}` is available with Lua 5.3 - and if we have it, it's probably that LuaJIT picks it from Lua5.3.

**Reply by Frenzie** [reply]
Some history.
`https://github.com/LuaJIT/LuaJIT/issues/19`
`https://github.com/LuaJIT/LuaJIT/pull/70/files`

I'll probably propose this simple diff upstream, but I'm afraid they might prefer something much more complex like xgettext does in JS and Python. And if so I hope they'll take care of it, but until the next release we'd need some kind of patched version regardless, give or take doing a sed/whatever `\u` → `\\u` *before* xgettext gets its hands on it.

```diff
diff --git a/gettext-tools/src/x-lua.c b/gettext-tools/src/x-lua.c
index 78f0d179b..4e0955ba4 100644
--- a/gettext-tools/src/x-lua.c
+++ b/gettext-tools/src/x-lua.c
@@ -635,6 +635,11 @@ phase3_get (token_ty *tp)
                     case 't':
                       string_add ('\t');
                       break;
+                    case 'u':
+                    case 'U':
+                      string_add ('\\');
+                      string_add (c);
+                      break;
                     case 'v':
                       string_add ('\v');
                       break;
```

**Reply by Frenzie** [reply]
Btw, I don't really know what the emoji font is, but the symbol seems to be broken.
<img src=https://github.com/koreader/koreader/assets/202757/d3076964-7a37-497a-9752-c0184f51135e width=40%>

**Reply by poire-z** [reply]
Seems it was already broken at the time #9583 - and I didn't even notice it while making the screenshot ?!!! :/

But ! It might be a feature ! Phew! :)

When `[x] display font name in their own fonts`, if the font doesn't have it, you get the broken symbol. If the font has it, you get it:
![image](https://github.com/koreader/koreader/assets/24273478/136a040a-8a04-462b-9431-1224c5b2988e)

When `[ ] display font name in their own fonts`, it is displayed with our main font + all our fallback fonts - but none of them include emojis - so you get the broken symbol. 

So, it's an indicator you don't have added and set a emoji'able font.
(I don't think we can programatically and easily know that to display something else...)

**Reply by Frenzie** [reply]
Actually keeping it as `\u` is problematic because we're reading from a file there. So the difficult way (in xgettext) is likely a necessity to keep things halfway sane.

**Reply by Frenzie** [reply]
Perhaps `\u` could remain relatively sane along such lines. NB Pseudocode, plus it'd probably be circular, not a fan of the idea.
```lua
local parsedContent = content:gsub("\\u{([%x]+)}", function(codePoint)
  local unicode = tonumber(codePoint, 16)
  return util.unicodeCodepointToUtf8(unicode)
end)
```

**Comment by ichnilatis-gr** [comment]
I couldn't imagine causing all this discussion, of which I understand nothing... :-)

**Reply by mergen3107** [reply]
Thanks for pointing this out!
